src/rest/applications/mcm/resubmission.py

```diff
@@ -0,0 +1,428 @@
+from enum import Enum
+from typing import Union
+
+from rest import McM
+from rest.applications.mcm.invalidate_request import InvalidateDeleteRequests
+from rest.utils.miscellaneous import pformat
+
+
+class RootRequestReset(Enum):
+    """
+    Describes how a root request was reset so that
+    it is reusable in later steps to recreate its
+    chained requests.
+    """
+
+    # Root request was soft reset.
+    SOFT_RESET = 1
+
+    # Root request was fully reset.
+    # This request was validated before, so
+    # instead of repeating the validation,
+    # old results are reincluded again.
+    FULL_RESET = 2
+
+    # Root request keeps output
+    # (it created a big "RAW" dataset being used in other steps, chains, ...)
+    # Handle it so that it is not required to recreate it
+    # again.
+    KEEPS_OUTPUT = 3
+
+
+class ChainRequestResubmitter:
+    """
+    Rewinds a chained request to its root
+    so that you can recreate the chain and attempt to
+    inject it again.
+
+    The scenario to use this is the following: Some configuration
+    applied in a campaign or chained campaign was wrong so that, after
+    applying a patch on these elements (that work as a template) you
+    need to recreate all the requests using this new change and reinject them.
+
+    Attributes:
+        mcm: McM client instance
+    """
+
+    def __init__(self, mcm: McM) -> None:
+        self._mcm = mcm
+        self._invalidator = InvalidateDeleteRequests(mcm=self._mcm)
+        self.logger = self._mcm.logger
+
+    def _root_request_to_approve(self, root_prepid: str) -> RootRequestReset:
+        """
+        This is a customization for resetting a root request so
+        that it ends up in an 'approve/approved' status and could
+        be easily reused to recreate a chained request and inject it
+        again.
+
+        Args:
+            root_prepid: Root request's prepid
+
+        Returns:
+            An attribute describing how the root request was processed.
+        """
+        data: Union[dict, None] = self._mcm.get(
+            object_type="requests", object_id=root_prepid
+        )
+        if not data:
+            raise ValueError(f"Root request {root_prepid} not found")
+
+        approval = data.get("approval")
+        status = data.get("status")
+        if approval == "submit" and status == "done":
+            keep_output: Union[list[bool], None] = data.get("keep_output")
+            if not keep_output:
+                raise ValueError("Keep output is not defined")
+
+            if all(keep_output):
+                # Avoid resetting this kind of root request
+                # Just reuse it as it is
```

Comment:
I don't understand the reasoning behind this... aren't the intermediate datasets also bad if we are resubmitting?

Reply:
They have been already deleted and invalidated. Details [here](https://github.com/cms-PdmV/mcm_scripts/blob/70d710f49a8d2dd5b5a7be6111311f124d96879f/src/rest/applications/mcm/resubmission.py#L393-L401).

Reply:
McM is so nice to work with :cold_sweat: